### ISSUE TITLE
False positive for `panicking_unwrap` on field in struct

### ISSUE THREAD

**Issue by mkeeter** [rust-lang/rust-clippy#16467]
### Summary

I noticed a strange CI failure after the `rustc 1.93` release:

https://github.com/mkeeter/fidget/actions/runs/21339812932/job/61417400107

This boiled down to the reproducer shown below, and appears to be a false positive in `panicking_unwrap`.

Note that the indirection of having the field in a `struct` is load-bearing for the bug; I couldn't reproduce the issue with an argument of `foo: &mut Option<u32>`.

### Lint Name

panicking_unwrap

### Reproducer

I tried this code:

```rust
struct Omg {
    bar: Option<u32>,
}

fn check_option(foo: &mut Omg) -> &u32 {
    if foo.bar.is_none() {
        foo.bar = Some(15);
        &foo.bar.as_ref().unwrap()
    } else {
        &foo.bar.as_ref().unwrap()
    }
}

fn main() {
    let mut a = Omg { bar: None };
    dbg!(check_option(&mut a));
    let mut b = Omg { bar: Some(500) };
    dbg!(check_option(&mut b));
}
```

I saw this happen:

```
error: this call to `unwrap()` will always panic
 --> src/main.rs:8:10
  |
6 |     if foo.bar.is_none() {
  |        ----------------- because of this check
7 |         foo.bar = Some(15);
8 |         &foo.bar.as_ref().unwrap()
  |          ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.93.0/index.html#panicking_unwrap
  = note: `#[deny(clippy::panicking_unwrap)]` on by default
```

I expected not to see this error, because it isn't true; indeed, running `main` does not panic.

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=b47329aac1fc27450ae035c2b3452cde)

### Version

```text
Running on the playground with `clippy 0.1.93 (2026-01-19 254b59607d)`
```

**Comment by theemathas** [comment]
This is a duplicate of #16436, I believe.

**Comment by flip1995** [comment]
_If_ there is a point release 1.93.1 we'll include the fix for this there. If not, this fix will be released with 1.94.0. Sorry for the inconvenience.
